staging/src/k8s.io/externaljwt/pkg/plugin/plugin.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"gopkg.in/square/go-jose.v2/jwt"
+	"k8s.io/externaljwt"
+	externaljwtv1alpha1 "k8s.io/externaljwt/apis/v1alpha1"
+)
+
+type VerificationKeys struct {
+	Keys          []externaljwt.PublicKey
+	DataTimestamp time.Time
+}
+
+// New calls external signer to fill out supported keys.
+// It also starts a periodic sync of external keys.
+// In order for the key cache and external signing to work correctly, pass a context that will live as
+// long as the dependnat proccess; is used to maintain the lifetime of the connection to external signer.
+func New(ctx context.Context, issuer, socketPath string, keySyncTimeout time.Duration, preventSigningWithNonOIDCKeys bool) (*Plugin, *keyCache, error) {
+	conn, err := grpc.Dial(
+		socketPath,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithAuthority("localhost"),
+		grpc.WithDefaultCallOptions(grpc.WaitForReady(true)),
```

Comment:
how long do we wait for ready before failing Dial / client construction? I think that will determine how long kube-apiserver blocks on start before either succeeding or failing

Reply:
WaitForReady waits forever but our first call (InitialFill) uses a context timeout of 60s.

I have added a test case to async start the signer 30s after kube-apiserver has started to initialize, seem to work.

Reply:
Can you manually confirm that we do eventually fail the API server start if the external signer never comes up?

Reply:
Done

tested by setting the initial fill timeout to 10 sec while delaying the mock signer to start 30s after kube-apiserver. It resulted in the test failing and failing early.

Reply:
> WaitForReady waits forever but our first call (InitialFill) uses a context timeout of 60s.
> ...
> tested by setting the initial fill timeout to 10 sec while delaying the mock signer to start 30s after kube-apiserver. It resulted in the test failing and failing early.

I'm a little confused by that ... does WaitForReady not wait in the Dial call? I thought that wait happened before it returned a `conn`.

Reply:
I was wrong about what WaitForReady does... it doesn't block Dial, it blocks the RPC call. That is the first FetchKeys call, which is done with a timeout context